demos/common/monitors/memory_monitor.cpp

```diff
@@ -0,0 +1,258 @@
+}
+
+double MemoryMonitor::getMeanMem() const {
+    return meanMem;
+}
+
+double MemoryMonitor::getMeanSwap() const {
+    return meanSwap;
+}
+
+double MemoryMonitor::getMaxMem() const {
+    return maxMem;
+}
+
+double MemoryMonitor::getMaxSwap() const {
+    return maxSwap;
+}
+#else
+#include <fstream>
+#include <utility>
+#include <vector>
+#include <regex>
+
+namespace {
+double getMemTotal() {
+    double memTotal = 0;
+    std::regex memRegex("^(.+):\\s+(\\d+) kB$");
+    std::string line;
+    std::smatch match;
+    std::ifstream meminfo("/proc/meminfo");
+    std::getline(meminfo, line);
+    while(meminfo.good())
+    {
+        if (std::regex_match(line, match, memRegex))
+        {
+            if ("MemTotal" == match[1]) {
+                memTotal = stod(match[2]) / (1024 * 1024);
+            }
+        }
+        std::getline(meminfo, line);
+    }
+    if (0 == memTotal) {
+        throw std::runtime_error("Can't get total memory");
+    } else {
+        return memTotal;
+    }
+}
+
+double getSwapTotal() {
+    double swapTotal = 0;
+    std::regex memRegex("^(.+):\\s+(\\d+) kB$");
+    std::string line;
+    std::smatch match;
+    std::ifstream meminfo("/proc/meminfo");
+    std::getline(meminfo, line);
+    while(meminfo.good())
+    {
+        if (std::regex_match(line, match, memRegex))
+        {
+            if ("SwapTotal" == match[1]) {
+                swapTotal = stod(match[2]) / (1024 * 1024);
+            }
+        }
+        std::getline(meminfo, line);
+    }
+    if (0 == swapTotal) {
+        throw std::runtime_error("Can't get total memory");
+    } else {
+        return swapTotal;
+    }
+}
+}
+
+std::pair<double, double> getAvailableMemSwap() {
+    double availableMem = 0, availableSwap = 0;
+    unsigned long memfree = 0, buffers = 0, cached = 0, sReclaimable = 0, shmem = 0;
+    std::regex memRegex("^(.+):\\s+(\\d+) kB$");
+    std::string line;
+    std::smatch match;
+    std::ifstream meminfo("/proc/meminfo");
+    std::getline(meminfo, line);
+    while(meminfo.good())
+    {
+        if (std::regex_match(line, match, memRegex))
+        {
+            if ("MemFree" == match[1]) {
+                memfree = stoul(match[2]);
+            } else if ("Buffers" == match[1]) {
+                buffers = stoul(match[2]);
+            } else if ("Cached" == match[1]) {
+                cached = stoul(match[2]);
+            } else if ("SReclaimable" == match[1]) {
+                sReclaimable = stoul(match[2]);
+            } else if ("Shmem" == match[1]) {
+                shmem = stoul(match[2]);
+            } else if ("SwapFree" == match[1]) {
+                availableSwap = stod(match[2]) / (1024 * 1024);
+            }
+        }
+        std::getline(meminfo, line);
+    }
+    if (0 == availableSwap) {
+        throw std::runtime_error("Can't get available swap memory");
+    }
+    availableMem = static_cast<double>(memfree + buffers + cached + sReclaimable - shmem) / (1024 * 1024);
```

Comment:
Why not just use `MemAvailable`?

Reply:
It follows `htop`'s implementation. I'm also not sure this is a good option because of comment in `man`:
> An estimate of how much memory is available for starting new applications, without swapping.

Reply:
> It follows htop's implementation.

Are you sure that implementation was not written before [MemAvailable was implemented](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=34e431b0ae398fc54ea69ff85ec700722c9da773)?

> > An estimate of how much memory is available for starting new applications, without swapping.

Isn't that exactly what we want?

Reply:
> Are you sure that implementation was not written before MemAvailable was implemented?

Looks that later: https://github.com/hishamhm/htop/commit/26422af6088129b99864b75ed4ce5395aa578f2b#diff-6a85ca50e4ed143060c0849894d89370R564, but it could be done for compatibility.

> Isn't that exactly what we want?

This note about "starting new applications" can mean that available memory for acquiring by existing applications and available memory for starting new applications aren't the same numbers. I can't say which we want for sure.

Reply:
> Looks that later:

The code you've shown me has actually [existed since before htop switched to Git](https://github.com/hishamhm/htop/commit/d6231bab89d634da5564491196b7c478db038505#diff-78718b9cfe8260b8487ef93a88226e44R490). Although I don't see where the available memory calculation that we are discussing is in this code.

> This note about "starting new applications" can mean that available memory for acquiring by existing applications and available memory for starting new applications aren't the same numbers.

Well... even if they aren't the same (which I find doubtful - why wouldn't they be?), isn't it still better to use the estimate explicitly provided by the kernel, rather than a formula of questionable origins?

Reply:
Verifying the numbers by running `htop` will give (almost) the same results. I suppose it is more convenient way of verifying and `htop` is the first thing that comes in mind. At least it was the first thing I wanted to check values with when I had an initial version.

To me `MemAvailable` gives different results (diff is about 0.4GiB).

Reply:
What are you comparing the numbers with? This meter?

![изображение](https://user-images.githubusercontent.com/42033526/68780041-ae4d4b80-0646-11ea-951d-a4fe5236c6b8.png)

Because that's calculated as [`used` - `buffers` - `cached`](https://github.com/hishamhm/htop/blob/04cc193e3c0c39ea47eb01d61a6866b32d70baea/linux/Platform.c#L203), which is not what you do either.

> I suppose it is more convenient way of verifying and htop is the first thing that comes in mind. At least it was the first thing I wanted to check values with when I had an initial version.

That's not much of an argument. I could tell you that the first thing that comes to my mind is `free` and that uses `MemAvailable`. :-)

Reply:
I am comparing to the number reported. It is 4.4G for you.

[The trick is how you get `cachedMem`](https://github.com/hishamhm/htop/blob/04cc193e3c0c39ea47eb01d61a6866b32d70baea/linux/LinuxProcessList.c#L960). `usedMem - buffersMem - cachedMem = usedMem - buffersMem - cachedMem(procfs) - sreclaimable + shmem`

> That's not much of an argument. I could tell you that the first thing that comes to my mind is free and that uses MemAvailable. :-)

OK than. I will update.

Reply:
updated